### ISSUE TITLE
no such module: fts5

### ISSUE THREAD

**Issue by alexeymorin** [wagtail/wagtail#7798]
### Issue Summary

site-packages\django\db\backends\sqlite3\base.py", line 421, in execute return Database.Cursor.execute(self, query)
django.db.utils.OperationalError: no such module: fts5

### Steps to Reproduce

1. Start a new project with `wagtail start myproject`
2. manage.py migrate

### Technical details

Python 3.7.6
Django 3.2.10
wagtail 2.15.1

**Comment by gasman** [comment]
I think this might be an issue with the version of Python in use - 3.7.6 is quite old (the latest release in the Python 3.7 branch is 3.7.12), and is most likely shipped with an outdated version of the sqlite library.

In the backend itself, we're careful to only import the FTS backend if the version is new enough https://github.com/wagtail/wagtail/blob/main/wagtail/search/backends/database/__init__.py#L16 , but there's no such check in the migration https://github.com/wagtail/wagtail/blob/main/wagtail/search/migrations/0006_customise_indexentry.py .

Updating to a newer Python release should fix this, but we should also fix the migration so that older versions are handled gracefully.

**Comment by kdkocev** [comment]
I have the same exact problem when I follow the installation guide on the official wagtail documentation.

Technical details

Python 3.9.9
Django 3.2.10
wagtail 2.15.1

@gasman Updating to a newer version of python doesn't to the trick here.
Is there any other way we can fix this without fiddling around with the migrations inside the wagtail package?

**Comment by mellowbro007** [comment]
I have the same problem and the same versions as @kdkocev.

**Comment by gasman** [comment]
Looking into this now. A couple of things I'll need to establish first:

* Is the failure of the line `CREATE VIRTUAL TABLE %s_fts USING fts5(autocomplete, body, title)` tied to specific sqlite versions, or is it some other system-dependent issue?
* What factors govern whether a given Python installation has an FTS-supporting version of sqlite? Does it depend on major Python release (evidently not, since our CI setup includes a test run against 3.7.12 and sqlite and didn't catch this), minor Python release (i.e. Python will routinely bump the bundled sqlite to a newer version even for 3.7.x patch releases), the whims of the person doing the packaging (so e.g. the Ubuntu package of Python 3.9.9 might potentially have a different sqlite version to the Windows installer for Python 3.9.9), or something else system-dependent?
* Is it expected that a user should be able to upgrade from a non-FTS version of sqlite to an FTS version without having to rebuild the database?

If the answer to the last question is yes, and non-FTS versions of sqlite do legitimately exist within our supported Python releases, then presumably the FTS index creation will have to happen outside of a migration. (Otherwise, even if we fix the migration to gracefully skip that step on sqlite versions that don't support it, it will break later on when the user upgrades to a newer version and the search backend attempts to use the non-existing index.)

**Comment by gasman** [comment]
From https://stackoverflow.com/questions/60246821/how-can-i-use-the-fts5-extension-with-the-sqlite3-python-module-with-python-3-7, it's clear that at least on Linux, it will pick up whatever version of libsqlite is installed system-wide. (For Windows and Mac, I suspect it's bundled as part of the installer.)

Judging from https://www.sqlite.org/fts5.html#compiling_and_using_fts5, it sounds like we can't use the sqlite version alone to determine whether fts5 is available either.

**Comment by gasman** [comment]
Suggested fix:

* Remove [the SQL operations that create the indexentry_fts virtual table](https://github.com/wagtail/wagtail/blob/main/wagtail/search/migrations/0006_customise_indexentry.py#L78-L81) from the `search.0006_customise_indexentry` migration
* Add [an 'is fts available' helper function](https://stackoverflow.com/a/36656216/1853523) to wagtail.search.backends.database
* Add a `post_migrate` signal handler to the wagtail.search app (the same thing [that django.contrib.contenttypes does](https://github.com/django/django/blob/63869ab1f191ab5781cde8b813b838300455f6d6/django/contrib/contenttypes/apps.py#L21) to populate the contenttypes table) that checks whether
  * the db is sqlite
  * fts5 is available
  * the indexentry_fts virtual table does not exist

  and if so, runs the SQL operations we removed from `search.0006_customise_indexentry`
* Update the version check in wagtail.search.backends.database.SearchBackend to use the fallback database backend if `sqlite_version_info < (3, 19, 0)` OR fts5 is unavailable

@kaedroho Does this plan sound OK?

**Comment by kaedroho** [comment]
@gasman I think this would be OK, just wondering if we need to think about:

 - Do we need a way to roll back the migration and delete the FTS table?
 - Would this make it difficult to change the FTS table later?

**Comment by ihaiderj** [comment]
It started working for me when I deleted the migration folder from the wagtail package (inside virtualenv in my case) and ran migrate command again and it worked

**Comment by ihaiderj** [comment]
> I have the same exact problem when I follow the installation guide on the official wagtail documentation.
> 
> Technical details
> 
> Python 3.9.9 Django 3.2.10 wagtail 2.15.1
> 
> Updating to a newer version of python doesn't to the trick here. @gasman Is there any other way we can fix this without fiddling around with the migrations inside the wagtail package?
> 
> The error doesn't exist if you run the same commands against a different database (postgresql for example), but I was wondering if there's a fix for sqlite3 since it's the default setting in the installation example.

It started working for me when I deleted the migration folder from the wagtail package (inside virtualenv in my case) and ran migrate command again and it worked

**Comment by gasman** [comment]
> It started working for me when I deleted the migration folder from the wagtail package (inside virtualenv in my case) and ran migrate command again and it worked

Don't do that. That'll make it incredibly difficult to upgrade to future Wagtail versions, because the migrations you generated and applied won't match the ones shipped with the next release of Wagtail.

**Comment by ihaiderj** [comment]
Thanks I ll take care

**Comment by ihaiderj** [comment]
Whats the best solution you suggest, mine was showing 11 migrations pending
after i ran migrate command.. and showing the cause  no such module: fts5

**Comment by gasman** [comment]
We're working on a proper fix. In the meantime, I'd recommend upgrading your installed version of sqlite to one that includes the fts5 extension - on Linux this will be a system package, so the process for upgrading will depend on your operating system. (In the case of Ubuntu, upgrading your system to 18.04LTS or later should work.)

Alternatively, you can switch to another database backend such as Postgresql - we'd recommend this anyhow once you start building 'real-world' production sites.

**Comment by ihaiderj** [comment]
Thanks, I'll check it out.

**Comment by gasman** [comment]
@kaedroho Good points. After some more pondering I'm slightly nervous about making schema changes outside of migrations, and having to permanently redesign things around an edge case that will hopefully disappear in a year or two. I think a better idea might be:

* keep the table creation in the migration, but make it conditional on checking that fts5 is available
* add an 'if table exists' check before dropping the table in the reverse migration
* On server startup, when the backend is initialised, we check that sqlite is new enough, fts5 is available and the table exists, and use the fallback backend if not. If sqlite does have usable fts5 support but the table doesn't exist, that means they've upgraded sqlite since running the migration, and in this case we should output a warning advising them that they should reverse and re-run migration 0006 to enable full-text search. (Or just give them instructions - or a management command? - to create the table themselves.)
* If any changes to the table are needed in future, they can go in subsequent migrations as normal, but will need an 'if table exists' check to accommodate users without fts5.

**Comment by gasman** [comment]
Work in progress at https://github.com/gasman/wagtail/tree/fix/sqlite-without-fts5.

> On server startup, when the backend is initialised, we check that sqlite is new enough, fts5 is available and the table exists

There's a subtle point about _when_ exactly we do this. For example, we can't put this check in AppConfig.ready, because that will run at the start of a `./manage.py migrate` command, when (obviously) the table won't exist on first run - so it would wrongly output the warning. Similarly, I don't think we can safely put the 'does table exist' check in any module-level code that gets imported via models.py, because we can't assume the database is ready by then.

As far as I can tell, it should be _mostly_ safe to put the 'does table exist' check [inside `wagtail.search.backends.database.SearchBackend`](https://github.com/wagtail/wagtail/blob/main/wagtail/search/backends/database/__init__.py#L15-L22), as that gets imported from `get_search_backend`, which will typically only get called when we actually want to perform a search operation. There's a possibility that there's some code out in the wild that calls `get_search_backend` at module load time and holds on to the backend instance to run a query when it feels like it, but if so, I think we just have to let that break.

In any case, we should cache the result of that check, so that repeated calls to get_search_backend() don't re-run the check.

**Comment by benwaah** [comment]
Hi,
The fix for this doesn't seem to be enough in my case.

Here are some tests I did :
`Python 3.8.6 (default, Sep  6 2021, 04:03:13)`
`[GCC 7.3.1 20180303 (Red Hat 7.3.1-5)] on linux`
`Type "help", "copyright", "credits" or "license" for more information.`
`>>> import sqlite3`
`>>> sqlite3.sqlite_version_info`
`(3, 26, 0)`
`>>> conn = sqlite3.connect('mydatabase')`
`>>> conn.enable_load_extension(True)`
`>>> conn.load_extension('fts5')`
`Traceback (most recent call last):`
`  File "<stdin>", line 1, in <module>`
`sqlite3.OperationalError: fts5.so: cannot open shared object file: No such file or directory`

How about a configuration flag to force wagtail to use the dummy fallback backend in this case?

**Comment by gasman** [comment]
@benwaah Can you give more detail about where it's failing now? The sqlite library throwing an `OperationalError` should be exactly the situation that we recognise, and respond to by enabling the fallback backend: https://github.com/wagtail/wagtail/blob/92a323231f10a0b990e0a14f8b4e37ab9dc3a4a3/wagtail/search/backends/database/sqlite/utils.py#L15-L18

**Comment by gasman** [comment]
Yes, and the code within Wagtail is catching that exception - I need to know what specific part of the Wagtail setup process is failing, rather than what equivalent errors you can trigger from the Python command line.

**Comment by benwaah** [comment]
For some reason the exception does not seem to be catched when I run `./manage.py migrate wagtailsearch 0005` for instance:

`Traceback (most recent call last):`
`[...]`
`File "venv/lib/python3.8/site-packages/wagtail/search/models.py", line 214, in <module>`
`    if fts5_available():`
`  File "venv/lib/python3.8/site-packages/wagtail/search/backends/database/sqlite/utils.py", line 16, in fts5_available`
`    tmp_db.execute('CREATE VIRTUAL TABLE fts5test USING fts5 (data);')`
`sqlite3.OperationalError: no such module: fts5`

**Comment by gasman** [comment]
ah - I guess we ought to be catching `sqlite3.OperationalError` rather than Django's own `OperationalError`...

**Comment by gasman** [comment]
At this point this is extremely difficult to replicate, since you need an environment that's old enough to not have sqlite-with-fts5 baked in, but new enough to evade the >3.19 version check, and new enough to let you install Python>=3.6 at all. The best I could do was to spin up an Ubuntu 16.04 instance in Vagrant and fake the sqlite version...

```
vagrant init ubuntu/xenial64
vagrant up
vagrant ssh
sudo apt update
sudo apt install -y python3 git curl build-essential libbz2-dev libssl-dev libreadline-dev libffi-dev libsqlite3-dev tk-dev
curl https://pyenv.run | bash
echo 'export PATH="$HOME/.pyenv/bin:$PATH"' >> ~/.bashrc
echo 'eval "$(pyenv init -)"' >> ~/.bashrc
echo 'eval "$(pyenv virtualenv-init -)"' >> ~/.bashrc
exec $SHELL
pyenv install 3.7.12
~/.pyenv/versions/3.7.12/bin/python -m venv ~/.venv/wagtail
source ~/.venv/wagtail/bin/activate
git clone https://github.com/wagtail/wagtail.git
cd wagtail
pip install -e .

# edit wagtail/search/backends/database/sqlite/utils.py and change (3, 19, 0) to (3, 1, 0)

cd ..
wagtail start myapp
cd myapp
./manage.py migrate
```

This does indeed fail with `sqlite3.OperationalError: no such module: fts5`, and updating the `fts5_available` function to catch `sqlite3.OperationalError` instead of `django.db.OperationalError` fixes it. PR incoming...

**Comment by gasman** [comment]
Fixed in #7893
